File: frontend/src/app/services/treats.service.ts
import {Injectable} from '@angular/core';
import {HttpClient, HttpParams} from '@angular/common/http';
import {Observable} from 'rxjs';
import {environment} from '../../environments/environment';
import {PatientRequest, Treats} from '../dtos/patient';

const baseUri = environment.backendUrl + '/api/v1/treats';

@Injectable({
  providedIn: 'root'
})
export class TreatsService {
  constructor(
    private http: HttpClient,
  ) {
  }

  /**
   * Get all requests for the logged in.
   *
   * @Retruns a list of requests from doctors to treat the logged in patient
   */
  getAllRequests(search: string): Observable<Treats[]> {
    const params = {params: new HttpParams().set('search', search)};
    return this.http.get<Treats[]>(baseUri + '/requests', params);
  }

  /**
   * Request a treating relationship to a patient (as a doctor).
   *
   * @param patientId the patient to treat
   */
  requestTreats(patientId: number): Observable<PatientRequest> {
    return this.http.post<PatientRequest>(baseUri + '/' + patientId, null);
  }

  /**
   * Respond to a request posed by the doctor
   *
   * @param doctorId the id of the doctor who sent the request
   * @param accept true: accept request, false: decline request
   */
  respondToRequest(doctorId: number, accept: boolean): Observable<Treats> {
    return this.http.put<Treats>(baseUri + '/' + doctorId, accept);
  }

  /**
   * Delete a treating relationship
   *
   * @param id of the person to delete the treating relationship with
   */
  deleteTreats(id: number): Observable<void> {
    return this.http.delete<void>(baseUri + '/' + id);
  }
}

File: backend/src/main/java/at/ac/tuwien/sepm/groupphase/backend/service/TreatsService.java
package at.ac.tuwien.sepm.groupphase.backend.service;

import at.ac.tuwien.sepm.groupphase.backend.endpoint.dto.PatientRequestDto;
import at.ac.tuwien.sepm.groupphase.backend.endpoint.dto.TreatsDto;
import at.ac.tuwien.sepm.groupphase.backend.entity.Doctor;
import at.ac.tuwien.sepm.groupphase.backend.entity.Patient;

import java.util.List;

/**
 * Service for managing treats relationships.
 */
public interface TreatsService {

    /**
     * Get all requests for a user (doctor or patient).
     *
     * @param userId the id of the user
     * @param search the search string
     * @return a list of all requests
     */
    List<TreatsDto> getAllRequests(long userId, String search);

    /**
     * Request a treats relationship with a patient.
     *
     * @param doctorId  the id of the doctor that wants to treat the patient
     * @param patientId the id of the patient that is requested
     * @return the created request
     */
    PatientRequestDto requestTreats(Long doctorId, Long patientId);

    /**
     * Respond to a treats request.
     *
     * @param patientUserId the id of the patient that is responding
     * @param doctorId      the id of the doctor that wants to treat the patient
     * @param accepted      whether the request was accepted or not
     * @return the updated treats relationship
     */
    TreatsDto respondToRequest(long patientUserId, Long doctorId, boolean accepted);

    /**
     * Create a treats relationship between a doctor and a patient.
     *
     * @param doctor  the doctor
     * @param patient the patient
     */
    void doctorTreatsPatient(Doctor doctor, Patient patient);

    /**
     * Delete a treats relationship between.
     *
     * @param sessionUserId the user id of the person wanting to delete the relationship
     * @param userId        the user id of the person to delete the relationship with
     */
    void deleteTreats(Long sessionUserId, Long userId);
}
